client.go

```diff
@@ -0,0 +1,354 @@
+	// return an error if any resource is invalid, and this error will be propagated as a NACK to the xDS
+	// backend.
+	Notify(resources iter.Seq2[string, *ads.Resource[T]]) error
+}
+
+// Watch functions much like [Cache.Subscribe]. The given [Watcher] will be notified whenever a
+// resource is updated. If the requested resource name is not already subscribed to, this will
+// trigger a new subscription for the client. If a resource is already known (for example from a
+// previous existing subscription), the watcher will be immediately notified.
+func Watch[T proto.Message](c *ADSClient, name string, watcher Watcher[T]) {
+	if getResourceHandler[T](c).AddWatcher(name, watcher) {
+		c.notifyNewSubscription()
+	}
+}
+
+// getResourceHandler gets or initializes the [internal.ResourceHandler] for the specified type in
+// the given client.
+func getResourceHandler[T proto.Message](c *ADSClient) *internal.ResourceHandler[T] {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	typeURL := utils.GetTypeURL[T]()
+	if hAny, ok := c.handlers[typeURL]; !ok {
+		h := internal.NewResourceHandler[T]()
+		c.handlers[typeURL] = h
+		return h
+	} else {
+		return hAny.(*internal.ResourceHandler[T])
+	}
+}
+
+func (c *ADSClient) getResourceHandler(typeURL string) (internal.RawResourceHandler, bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	h, ok := c.handlers[typeURL]
+	return h, ok
+}
+
+// notifyNewSubscription signals to the subscription loop that a new subscription was added.
+func (c *ADSClient) notifyNewSubscription() {
+	select {
+	case c.newSubscription <- struct{}{}:
+	default:
+	}
+}
+
+type subscriptionSet map[string]utils.Set[string]
+
+// getPendingSubscriptions iterates over all the subscriptions returned by invoking
+// [internal.ResourceHandler.AllSubscriptions] on all registered resource handlers, and compares it
+// against the given set of already registered subscriptions. If any are missing, they are added to
+// the returned subscription set after being added to the given set. This means that repeated
+// invocations of this method will return an empty set if no new subscriptions are added in between.
+func (c *ADSClient) getPendingSubscriptions(registeredSubscriptions subscriptionSet) subscriptionSet {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	pendingSubscriptions := make(subscriptionSet)
+	add := func(typeURL string, name string) {
+		registered := internal.GetNestedMap(registeredSubscriptions, typeURL)
+		if !registered.Contains(name) {
+			registered.Add(name)
+			internal.GetNestedMap(pendingSubscriptions, typeURL).Add(name)
+		}
+	}
+
+	for t, handler := range c.handlers {
+		for k := range handler.AllSubscriptions() {
+			add(t, k)
+		}
+	}
+
+	return pendingSubscriptions
+}
+
+// loop simply calls newStream and subscriptionLoop forever, until the underlying gRPC connection is
+// closed.
+func (c *ADSClient) loop() {
+	for {
+		stream, responses, err := c.newStream()
```

Comment:
could use some comment here? such as subscriptionLoop below won't exit until the previous stream is closed, so we need to create a new stream on each iteration.

Reply:
Sure, I can add that